homeassistant/components/enphase_envoy/sensor.py

```diff
@@ -136,33 +144,102 @@ def icon(self):
     @property
     def device_state_attributes(self):
         """Return the state attributes."""
-        if self._type == "inverters":
-            return {"last_reported": self._last_reported}
+        if self.sensor_type == "inverters":
+            return {"last_reported": self.last_reported}
 
         return None
 
     async def async_update(self):
-        """Get the energy production data from the Enphase Envoy."""
-        if self._type != "inverters":
-            _state = await getattr(self._envoy_reader, self._type)()
-            if isinstance(_state, int):
-                self._state = _state
-            else:
-                _LOGGER.error(_state)
-                self._state = None
+        """Get the latest data from the Enphase Envoy device."""
+        await self.envoy_data.update()
+        self.envoy_data.set_state(self)
+        self.envoy_data.set_entity_attributes(self)
 
-        elif self._type == "inverters":
-            try:
-                inverters = await (self._envoy_reader.inverters_production())
-            except requests.exceptions.HTTPError:
-                _LOGGER.warning(
-                    "Authentication for Inverter data failed during update: %s",
-                    self._envoy_reader.host,
-                )
 
-            if isinstance(inverters, dict):
-                serial_number = self._name.split(" ")[2]
-                self._state = inverters[serial_number][0]
-                self._last_reported = inverters[serial_number][1]
+class EnvoyData:
```

Comment:
Any reason you went for your own class instead of using the DataUpdateCoordinator? https://developers.home-assistant.io/docs/integration_fetching_data